### PR TITLE
IOS-168: Implement Privacy & Safety Settings

### PR DESCRIPTION
**Pull request by kimar**
| Rationale | Demo |
|---|---|
| Implements a settings screen to control your Privacy & Safety settings from within the App. | RocketSim_Recording_iPhone_15_6 1_2024-05-30_11 58 40 |

### CODE REVIEW

File: MastodonSDK/Sources/MastodonSDK/Entity/Mastodon+Entity+Account.swift [hunk 1]
```diff
@@ -37,6 +37,8 @@ extension Mastodon.Entity {
         public let locked: Bool
         public let emojis: [Emoji]
         public let discoverable: Bool?
+        public let hideCollections: Bool?
+        public let indexable: Bool?
```
Comment (zeitschlag) [inline MastodonSDK/Sources/MastodonSDK/Entity/Mastodon+Entity+Account.swift]: If I have a look at https://docs.joinmastodon.org/entities/account/, I don't find these values. Where do you get these from?

File: Mastodon/Scene/Settings/Privacy and Safety/PrivacySafetyViewModel.swift [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+            preset = .openPublic
+        } else if self == Self.privateRestricted {
+            preset = .privateRestricted
+        } else {
+            preset = .custom
+        }
+    }
+
+    func loadSettings() {
+        Task { @MainActor in
```
Comment (zeitschlag) [inline Mastodon/Scene/Settings/Privacy and Safety/PrivacySafetyViewModel.swift:140]: Just out of curiosity: Is there a difference between `Task { @MainActor in` and the `@MainActor`-attribute?

File: Mastodon/Scene/Settings/Privacy and Safety/PrivacySafetyViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+                domain: domain,
+                query: .init(
+                    discoverable: suggestMyAccountToOthers,
+                    locked: manuallyApproveFollowRequests,
+                    source: .withPrivacy(visibility.toPrivacy()),
+                    indexable: appearInSearches,
+                    hideCollections: !showFollowersAndFollowing
+                ),
+                authorization: userAuthorization
+            ).value
```
Comment (zeitschlag) [inline Mastodon/Scene/Settings/Privacy and Safety/PrivacySafetyViewModel.swift]: What do we need this `.value` for?

File: Mastodon/Scene/Settings/Privacy and Safety/PrivacySafetyViewController.swift [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        title = L10n.Scene.Settings.PrivacySafety.title
+    }
+}
+
+struct PrivacySafetyView: View {
```
Comment (zeitschlag) [inline Mastodon/Scene/Settings/Privacy and Safety/PrivacySafetyViewController.swift]: I was a bit surprised to see this `View` (and the `CheckableButton`) in the same file as the `PrivacySafetyViewController`, would it make sense for you to move them into their own files?

File: Mastodon/Scene/Settings/Privacy and Safety/PrivacySafetyViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        vm.visibility = .public
+        vm.manuallyApproveFollowRequests = false
+        vm.showFollowersAndFollowing = true
+        vm.suggestMyAccountToOthers = true
+        vm.appearInSearches = true
+        return vm
+    }()
+    
+    static let privateRestricted: PrivacySafetyViewModel = {
+        let vm = PrivacySafetyViewModel(appContext: nil, authContext: nil, coordinator: nil)
```
Comment (zeitschlag) [inline Mastodon/Scene/Settings/Privacy and Safety/PrivacySafetyViewModel.swift]: I needed a moment to understand, why the parameters are nil, but from my understanding, `openPublic` and `privateRestricted` both serve as templates, right? Their `appContext` etc. is totally irrelevant, as their other values get applied to a `PrivacySafetyViewModel`, which already has a `appContext, ...`, right?

**Review by zeitschlag — APPROVED** [review]
Works like a charm 👍 

As always: Some (more or less dumb) questions with one exception: Where do these new properties in `Mastodon.Entity.Account` come from? They don't show up in the docs

File: Mastodon/Scene/Settings/Privacy and Safety/PrivacySafetyViewModel.swift (hunk 2 above)
Comment (DK1DA) [reply Mastodon/Scene/Settings/Privacy and Safety/PrivacySafetyViewModel.swift:140]: I wanted to use a `Task` in this place to not have the ` loadSettings` function have to be invoked in an async context. But other than that i think it shouldn't make a difference.

File: Mastodon/Scene/Settings/Privacy and Safety/PrivacySafetyViewModel.swift (hunk 3 above)
Comment (DK1DA) [reply Mastodon/Scene/Settings/Privacy and Safety/PrivacySafetyViewModel.swift]: Oh I think that's a debug left-over, will remove it.

File: Mastodon/Scene/Settings/Privacy and Safety/PrivacySafetyViewController.swift (hunk 4 above)
Comment (DK1DA) [reply Mastodon/Scene/Settings/Privacy and Safety/PrivacySafetyViewController.swift]: Yes I'm doing this 👍

File: Mastodon/Scene/Settings/Privacy and Safety/PrivacySafetyViewModel.swift (hunk 5 above)
Comment (DK1DA) [reply Mastodon/Scene/Settings/Privacy and Safety/PrivacySafetyViewModel.swift]: Yes these defintions are only used for comparing agains the current settings and applying presets, so the context here is not consumed. Now that I'm writing this, I think the cleaner solution would be to use protocols for this, I'll fix this.

Comment (DK1DA) [reply Mastodon/Scene/Settings/Privacy and Safety/PrivacySafetyViewModel.swift]: And it's done, thanks for the Feedback!

File: Mastodon/Scene/Settings/Privacy and Safety/PrivacySafetyViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+        vm.visibility = .public
+        vm.manuallyApproveFollowRequests = false
+        vm.showFollowersAndFollowing = true
+        vm.suggestMyAccountToOthers = true
+        vm.appearInSearches = true
+        return vm
+    }()
+    
+    static let privateRestricted: PrivacySafetyViewModel = {
```
Comment (zeitschlag) [inline Mastodon/Scene/Settings/Privacy and Safety/PrivacySafetyViewModel.swift]: I wondered why you used the protocol above, but here the ViewModel?

```suggestion
    static let privateRestricted: PrivacySafetySettingApplicable = {
```

**Review by zeitschlag — APPROVED** [review]
👍

File: Mastodon/Scene/Settings/Privacy and Safety/PrivacySafetyViewModel.swift (hunk 6 above)
Comment (DK1DA) [reply Mastodon/Scene/Settings/Privacy and Safety/PrivacySafetyViewModel.swift]: That was actually a leftover this whole code, except from the `apply(from:)` method, can go . I'll remove it.

**Pull request merged**